src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectCreationExpression.cs

```diff
@@ -124,29 +131,80 @@ public override BoundNode VisitWithExpression(BoundWithExpression withExpr)
             //      tmp.P2 = e2;
             //      tmp
 
-            BoundExpression expression;
+            BoundExpression rewrittenReceiver = VisitExpression(receiver);
 
-            if (withExpr.Type.IsValueType)
+            if (type.IsAnonymousType)
             {
-                expression = VisitExpression(withExpr.Receiver);
+                var anonymousType = (AnonymousTypeManager.AnonymousTypePublicSymbol)type;
+                var sideEffects = ArrayBuilder<BoundExpression>.GetInstance();
+                var temps = ArrayBuilder<LocalSymbol>.GetInstance();
+                BoundLocal oldValue = _factory.StoreToTemp(rewrittenReceiver, out BoundAssignmentOperator boundAssignmentToTemp);
+                temps.Add(oldValue.LocalSymbol);
+                sideEffects.Add(boundAssignmentToTemp);
+
+                BoundExpression value = _factory.New(anonymousType, getAnonymousTypeValues(withExpr, oldValue, anonymousType, sideEffects, temps));
+
+                return new BoundSequence(withExpr.Syntax, temps.ToImmutableAndFree(), sideEffects.ToImmutableAndFree(), value, type);
+            }
+
+            BoundExpression expression;
+            if (type.IsValueType)
+            {
+                expression = rewrittenReceiver;
             }
             else
             {
                 Debug.Assert(withExpr.CloneMethod is not null);
                 Debug.Assert(withExpr.CloneMethod.ParameterCount == 0);
 
                 expression = _factory.Convert(
-                    withExpr.Type,
+                    type,
                     _factory.Call(
-                        VisitExpression(withExpr.Receiver),
+                        rewrittenReceiver,
                         withExpr.CloneMethod));
             }
 
             return MakeExpressionWithInitializer(
                 withExpr.Syntax,
                 expression,
                 withExpr.InitializerExpression,
-                withExpr.Type);
+                type);
+
+            ImmutableArray<BoundExpression> getAnonymousTypeValues(BoundWithExpression withExpr, BoundExpression oldValue, AnonymousTypeManager.AnonymousTypePublicSymbol anonymousType,
+                ArrayBuilder<BoundExpression> sideEffects, ArrayBuilder<LocalSymbol> temps)
+            {
+                var dictionary = PooledDictionary<PropertySymbol, BoundExpression>.GetInstance();
+                foreach (BoundExpression initializer in withExpr.InitializerExpression.Initializers)
+                {
+                    var assignment = (BoundAssignmentOperator)initializer;
```

Comment:
>(BoundAssignmentOperator)initializer; [](http://example.com/codeflow?start=37&length=37)

Is it that simple? It feels like we should be prepared at least to deal with some of the additional things that ```AddObjectInitializer``` helper is dealing with.